pebblo/app/service/prompt_service.py

```diff
@@ -0,0 +1,200 @@
+
+from pebblo.app.enums.enums import CacheDir
+from pebblo.app.libs.logger import logger
+from pebblo.app.models.models import AiDataModel, RetrievalContext, RetrievalData
+from pebblo.app.utils.utils import (
+    acquire_lock,
+    fields_validator,
+    read_json_file,
+    release_lock,
+    write_json_to_file,
+)
+
+
+class Prompt:
+    """
+    This class handles prompt API business logic.
+    """
+
+    def __init__(self, data: dict):
+        self.data = data
+        self.application_name = self.data.get("name")
+
+    def _fetch_retrieval_context_data(self):
+        """
+        Retrieve retrieval context from input data and return its corresponding model object.
+        """
+        logger.debug("Retrieving retrieval context details from input data")
+        # Fetching retrieval context details
+        retrieval_context_output = []
+        context_data = self.data.get("context")
+        if context_data:
+            for data in context_data:
+                fields_validator(data, ["retrieved_from", "doc", "vector_db"])
+                retrieval_context = RetrievalContext(
+                    retrieved_from=data.get("retrieved_from"),
+                    doc=data.get("doc"),
+                    vector_db=data.get("vector_db"),
+                )
+                retrieval_context_output.append(retrieval_context.dict())
+
+        logger.debug(
+            f"AI_APPS [{self.application_name}]: Retrieval Context Details: {retrieval_context_output}"
+        )
+        return retrieval_context_output
+
+    def _fetch_context_data(self, param):
+        """
+        Retrieve prompt/response data from input data and return its corresponding model object.
+        """
+        logger.debug(f"Retrieving {param} details from input data")
+
+        data = {}
+        context_data = self.data.get(param)
+        if context_data:
+            fields_validator(context_data, ["data"])
+            data = AiDataModel(
+                data=context_data.get("data"),
+                entityCount=context_data.get("entityCount")
+                if context_data.get("entityCount")
+                else 0,
+                entities=context_data.get("entities")
+                if context_data.get("entities")
+                else {},
+                topicCount=context_data.get("topicCount")
+                if context_data.get("topicCount")
+                else 0,
+                topics=context_data.get("topics") if context_data.get("topics") else {},
+            )
+
+        logger.debug(
+            f"AI_APPS [{self.application_name}]: {param} Details: {data.dict()}"
+        )
+        return data
+
+    def _create_retrieval_data(
+        self, retrieval_context_data, prompt_data, response_data
+    ):
+        """
+        Create an RetrievalData Model and return the corresponding model object
+        """
+        logger.debug("Creating RetrievalData model")
+        fields_validator(self.data, ["prompt_time", "user"])
```

Comment:
This function is not validating the input data, we need to return proper response here

Reply:
Okay, so If you validate input using this function, what response (message & status code) is returned if a particular field is not present in the input data?

Reply:
```
{
    "detail": "prompt_time is a mandatory field"
}
```
With 500 error

Reply:
Validation errors in the request body/data should fall under client-side errors (4xx) rather than Internal Server Error (500), right?
Also, this validation function will display an error for a single field even if two or more fields are missing in the request body.

Reply:
We can take this as a improvement and will fix in next iteration, creating issue for this https://github.com/daxa-ai/pebblo/issues/372